modules/api/src/main/java/org/apache/ignite/lang/CancelHandleImpl.java

```diff
@@ -0,0 +1,150 @@
+            if (t != null) {
+                cancelFut.completeExceptionally(t);
+            } else {
+                cancelFut.complete(null);
+            }
+        });
+    }
+
+    static final class CancellationTokenImpl implements CancellationToken {
+
+        private final CancelHandleImpl handle;
+
+        // Actions that trigger cancellations - action that triggers a cancellation + a future that completes, when a resource is closed.
+        private final ConcurrentLinkedQueue<Cancellation> cancellations = new ConcurrentLinkedQueue<>();
+
+        private final AtomicReference<CompletableFuture<Void>> cancelFutRef = new AtomicReference<>();
+
+        CancellationTokenImpl(CancelHandleImpl handle) {
+            this.handle = handle;
+        }
+
+        boolean isCancelled() {
+            return cancelFutRef.get() != null;
+        }
+
+        void addCancelAction(Runnable cancelAction, CompletableFuture<?> completionFut) {
+            assert cancelAction != null : "cancelAction must be not null";
+            assert completionFut != null : "completionFut must be not null";
+
+            Cancellation cancellation = new Cancellation(cancelAction, completionFut);
+            if (handle.isCancelled()) {
+                cancellation.run();
+            } else {
+                cancellations.add(cancellation);
+            }
+        }
+
+        @SuppressWarnings("rawtypes")
+        private CompletableFuture<Void> cancel() {
+            CompletableFuture<Void> f = cancelFutRef.get();
+            if (f != null) {
+                return f;
+            }
+
+            // First assemble all completion futures
+            CompletableFuture[] futures = cancellations.stream()
+                    .map(c -> c.completionFut)
+                    .toArray(CompletableFuture[]::new);
+
+            CompletableFuture<Void> cancelAll = CompletableFuture.allOf(futures);
+            CompletableFuture<Void> fut = cancelFutRef.compareAndExchange(null, cancelAll);
+
+            // If cancel future has not been set, trigger cancellation actions.
+            if (fut == null) {
+                for (Cancellation cancellation : cancellations) {
+                    cancellation.run();
+                }
+            }
+
+            return fut != null ? fut : cancelAll;
+        }
+    }
+
+    /**
+     * Stores an action that triggers a cancellation and a completable future that completes when a resource is closed.
+     */
+    private static class Cancellation {
+
+        private final Runnable cancelAction;
+
+        private final CompletableFuture<?> completionFut;
+
+        private Cancellation(Runnable cancelAction, CompletableFuture<?> completionFut) {
+            this.cancelAction = cancelAction;
+            this.completionFut = completionFut;
+        }
+
+        private void run() {
+            try {
+                cancelAction.run();
+            } catch (Throwable ignore) {
+                // Ignore
```

Comment:
If we have bugs in any cancelation action that will throw an exception, how can we know about the exception?

Reply:
Fixed.